src/main/java/com/google/javascript/clutz/DeclarationGenerator.java

```diff
@@ -69,6 +69,14 @@
   public static final Pattern GOOG_MODULE_EXTRACT =
       Pattern.compile("goog.module\\(['\"](.*)['\"]\\);");
 
+  /**
+   * Closure types can be used but not provided. Clutz collects them and
+   * visits them in a separate pass. A new set of types can be discovered
+   * while visiting unprovided types, thus recusively iterating. This holds
+   * the maximum depth the tool will visit.
+   */
+  private int maxTypeUsedDepth = 5;
```

Comment:
I thought java guys accepted the practical usefulness of mutable state :)

In all seriousness, indeed the pattern of collecting types and reemitting later stinks (hence the bugs I just caught). However, it is already in our codebase, so I want to push for this PR as a bug fix. Rolled in the two new fields into function local variables, so the opportunity for abuse is a lot less.

Switching to emitting all symbols is worth investigating and will clean up this particular issue, but might bring some other unknown problems. Issue #245